### PR TITLE
Change params.reads to params.input

### PR DESCRIPTION
**Pull request by drpatelh**
Many thanks for contributing to nf-core/tools!

Please fill in the appropriate checklist below (delete whatever is not relevant). These are the most common things requested on pull requests (PRs).

## PR checklist
 - [ ] This comment contains a description of changes (with reason)
 - [ ] If you've fixed a bug or added code that should be tested, add tests!
 - [ ] Documentation in `docs` is updated
 - [x] `CHANGELOG.md` is updated
 - [x] `README.md` is updated

**Learn more about contributing:** https://github.com/nf-core/tools/tree/master/.github/CONTRIBUTING.md

* Fixed bug in `markdown_to_html.py` to work with Python 2 and 3
* Make `params.input` mandatory and remove warnings for `params.single_end` and `params.design`
* Fixes https://github.com/nf-core/tools/issues/380

### CODE REVIEW

**Comment by ewels** [comment]
We've discussed switching `params.reads` for `params.input` before - although I see that it makes pipelines more similar to each other, I'm not super keen as it will break the workflows for a lot of existing users for relatively little gain..

Is it not sufficient to just allow either `params.input` or `params.reads` as I did in a recent PR to the linting?

**Comment by drpatelh** [comment]
I think its worth the effort to standardise this because it is the most basic parameter requirement for any given pipeline i.e. an input of some sort. Breaking workflows is not a bad thing in my opinion because it will force users to look at the new documentation/parameters between releases!

We are moving more towards the nf-core template becoming of general utility other than in NGS analysis and its possibly the right time that we have a provision for this? `reads` becomes redundant with `input` and I dont see any point in checking/having both...probably something that we should enforce sooner rather than later though due to your first point :sweat_smile:

**Comment by drpatelh** [comment]
I think its worth the effort to standardise this because it is the most basic parameter requirement for any given pipeline i.e. an input of some sort. Breaking workflows is not a bad thing in my opinion because it will force users to look at the new documentation/parameters between releases!

We are moving more towards the nf-core template becoming of general utility other than in NGS analysis and its possibly the right time that we have a provision for this? `reads` becomes redundant with `input` and I dont see any point in checking/having both...probably something that we should enforce sooner rather than later though due to your first point :sweat_smile:

**Comment by MaxUlysse** [comment]
I agree with @drpatelh we should switch to --input already.
`--reads` made sense for pipelines that can work directly from fastq files, but `--input` is more general, and works for everything.

We did break some stuff by switching to `snake_case` params already.

In sarek I do handle old params for retro-compatibility:
https://github.com/nf-core/sarek/blob/24a214e4f08ce2579011e1b36aeb2ae682877a62/main.nf#L122-L310

If you're afraid to break stuff, we could implement that.

**Comment by ewels** [comment]
I think I'm partially reticent about this because I like, _just_ rewrote a whole load of the linting code to allow the "one of several options" specifically for the `--reads` and `--input` case.

Personally I'm still slightly against this as I think that `--input` is less descriptive and intuitive than `--reads`. Standardising is good, but I'm more hesitant about it when the same flag across multiple pipelines does different things, as would be the case here (different input types).

But yeah, this is just a personal preference thing, so I'm quite happy to be outvoted 😄

**Comment by MaxUlysse** [comment]
How is `input` less descriptive and intuitive than `reads`?
:confused: 
But I do agree that `input` can be a lot of things, in sarek already input can be tsv files (paired fastq or recalibrated bam), a directory with paired end fastqs or even vcfs.
But for me, that even more a reason to use it than the contrary.

**Comment by drpatelh** [comment]
Yeah. I agree with @MaxUlysse. I think having a single parameter for pipeline input is the way to go. Im working on multiple pipelines now where `--reads` isnt an option along with `--single_end` (also removed here) and its a bit annoying getting the lint warnings...It can indeed be lots of different things but thats what the documentation is for...and multiple pipelines have already been using `--input` already without any confusion so I guess thats proof that it should be fine.

**Comment by drpatelh** [comment]
One bone of contention is that I have made `--input` mandatory here with the assumption that every pipeline will have to have some sort of input data! I think thats a valid assumption and will help to port everyone to use this after the next sync.

**Comment by Zethson** [comment]
While I can certainly understand the appeal of standardizing input parameters I just feel like it gets weird for pipelines with **multiple mandatory** parameters. Yes - sometimes there may be "main" input parameters like _reads_ or _spectra_ ,  but sometimes not. And naming just gets weird when you have 3+ mandatory parameters. Should they then have their names like they have now (params.database, params.design) or params.input_database and params.design? If their name is just 'input' it's not necessarily clear what the input actually should be. 'reads' or 'spectra' is very descriptive.

I am kind of split when it comes to this change.

**Comment by ewels** [comment]
I’m not saying that `--reads` is particularly better than `--input`, just that I don’t think it’s much worse. If you have to read the documentation anyway, why force standardisation on an input that isn’t standard? This is in contrast to other params (such as `--max_memory` or whatever) which do the exact same thing in each pipeline that uses them.

**Comment by drpatelh** [comment]
I think we also need to factor in the practicality of having `--reads` over `--input` in the pipeline template. For pipelines that don't use `--reads` these parameter changes have to be made in multiple places to remove it in order to use other developer defined equivalents so why not just use a more broadly applicable substitute? It's easy if you know you're way around the template but not for beginners. Yes, this will become easier with the `parameters.settings.json` implementation but it still means fewer changes will be required. Also, the lint warnings generated by `--reads` is often a source of confusion for new developers.

I agree that things are slightly more complicated when you have multiple mandatory parameters as input but in the majority of those cases there will still be a primary "raw" data input source. For example, in the `imcyto` pipeline I have used `--input` for the provision of multiple image file formats (something `--input` is also flexible to) and `--metadata` to provide a manifest type file - both are mandatory. So I guess there will be instances where this is possible but you just have to be sensible as to how you name things!
https://github.com/nf-core/imcyto/blob/6e210ca5cdca644f91049c40c52a04b2c661b849/main.nf#L30

If you bring me back here to provide my `--input` one more time...I WILL find the nearest bridge...`--reads` into this as you will.

**Comment by MaxUlysse** [comment]
> Yeah. I agree with @MaxUlysse. I think having a single parameter for pipeline input is the way to go. Im working on multiple pipelines now where `--reads` isnt an option along with `--single_end` (also removed here) and its a bit annoying getting the lint warnings...

I just added `params.single_end = false` to sarek to get rid of this annoying warning.
Thanks a lot for removing that as well.

**Comment by MaxUlysse** [comment]
> While I can certainly understand the appeal of standardizing input parameters I just feel like it gets weird for pipelines with **multiple mandatory** parameters. Yes - sometimes there may be "main" input parameters like _reads_ or _spectra_ , but sometimes not. And naming just gets weird when you have 3+ mandatory parameters. Should they then have their names like they have now (_params.database_, _params.design_) or _params.input_database_ and _params.design_? If their name is just _input_ it's not necessarily clear what the input actually should be. _reads_ or _spectra_ is very descriptive.
> 
> I am kind of split when it comes to this change.

I can totally see your point.
If you have several data type as an entry, why not making a tsv/csv file? and use that as an input then?

**Comment by drpatelh** [comment]
Whats the compromise here guys? Tell me and Ill make the changes 👍

**Comment by drpatelh** [comment]
gif

**Comment by ewels** [comment]
The compromise for me here would be to allow `--reads` in the linting but switch to `--input` in the template. Then we're encouraging everyone to use `--input` for standardisation but not forcing existing pipelines to switch.

**Comment by drpatelh** [comment]
Ok. I am going to start from scratch with this then because I cant be bothered to deal with the merge conflicts and dealing with the lint numbers etc. So just to clarify `--reads` and `--input` are both optional?

**Comment by apeltzer** [comment]
Yup, I guess thats the current solution. Update template once with `--input` and also allow`--reads`. Maybe enforce that one of them has to be present? Making both optional might end up having users without any of the two 😓

**Comment by ewels** [comment]
Currently we check that at least one of `params.reads`, `params.input` or `params.design` is present and throw a warning if not (see #542).

I think that a warning is probably fine? Having the template come with `--input` is a pretty strong nudge already, and every single PR will start showing the warning soon as a comment. So the level of annoyance should be pretty high 😁

**Comment by ewels** [comment]
ps. My resistance to this change is reducing more and more over time. I still kind of agree with https://github.com/nf-core/tools/pull/556#issuecomment-593131709 about parameters being descriptive, but I'm also quite happy to switch to a hard hard fail unless `params.input` if that's what most people want..

File: nf_core/lint.py [hunk 1]
```diff
@@ -429,16 +430,13 @@ def check_nextflow_config(self):
...
-            ['process.container'],
-            ['params.single_end']
+            ['process.container']
         ]
         # Old depreciated vars - fail if present
         config_fail_ifdefined = [
             'params.version',
             'params.nf_required_version',
             'params.container',
-            'params.singleEnd',
```
Comment (ewels) [inline nf_core/lint.py:441]: Why are you removing this? I think it's still good to fail if we come across the old-style of the parameter, as it suggests that the template hasn't been updated properly. Even if we are no long warning about the absence of the new-style.

File: nf_core/pipeline-template/{{cookiecutter.name_noslash}}/main.nf [hunk 2]
```diff
@@ -100,24 +100,24 @@ ch_output_docs_images = file("$baseDir/docs/images/", checkIfExists: true)
 /*
  * Create a channel for input read files
  */
-if (params.readPaths) {
+if (params.input_paths) {
```
Comment (ewels) [inline nf_core/pipeline-template/{{cookiecutter.name_noslash}}/main.nf:103]: Ugh, all this code is nasty 😞 

One day... https://github.com/nf-core/tools/issues/211

File: nf_core/lint.py (hunk 1 above)
Comment (drpatelh) [reply nf_core/lint.py]: Added back in: https://github.com/nf-core/tools/pull/556/commits/d1252c22b53eb380a5aee313f96de7d3834792d7

**Comment by drpatelh** [comment]
Thanks @ewels ! Ok, so at the moment:

* `--input`: mandatory so linting will fail without it
* `--reads`: non-mandatory and no warnings will be generated by linting if it isn't present

Everyone in agreement here? Speak now or forever hold you peas!

File: nf_core/pipeline-template/{{cookiecutter.name_noslash}}/main.nf (hunk 2 above)
Comment (drpatelh) [reply nf_core/pipeline-template/{{cookiecutter.name_noslash}}/main.nf:103]: Yeah, I know... I bypass it all using design files as `--input` 😆 Not nice though, agreed.

**Pull request merged**